openldap/2.6/debian-11/rootfs/opt/bitnami/scripts/libopenldap.sh

```diff
@@ -0,0 +1,1045 @@
+#   $2 - sleep between retries (in seconds). Default: 1
+# Returns:
+#   None
+#########################
+ldap_stop() {
+    local -r retries="${1:-12}"
+    local -r sleep_time="${2:-1}"
+
+    are_db_files_locked() {
+        local return_value=0
+        read -r -a db_files <<< "$(debug_execute find "$LDAP_DATA_DIR" -type f -print0 | xargs -0)"
+        for f in "${db_files[@]}"; do
+            result=$(fuser "$f" 2>&1)
+            pid=$(echo $result | cut -d ':' -f 2-)
+            if [ -n "$pid" ]; then
+                if is_boolean_yes "${SYMAS_DEBUG:-false}" ; then
+                    process=$(ps -p $"pid" -o comm=)
+                    warn "ldap_stop waiting on ${process} to close ${f} file"
+                fi
+                return_value=1
+            fi
+        done
+        return "$return_value"
+    }
+
+    is_ldap_not_running && return
+
+    stop_service_using_pid "$LDAP_PID_FILE"
+    if ! retry_while are_db_files_locked "$retries" "$sleep_time"; then
+        error "OpenLDAP failed to stop"
+        return 1
+    fi
+}
+
+########################
+# Create slapd.ldif
+# Globals:
+#   LDAP_*
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+ldap_create_slapd_file() {
+    info "Creating slapd.ldif"
+    cat > "${LDAP_SHARE_DIR}/slapd.ldif" << EOF
+#
+# See slapd-config(5) for details on configuration options.
+# This file should NOT be world readable.
+#
+
+dn: cn=config
+objectClass: olcGlobal
+cn: config
+olcArgsFile: ${LDAP_BASE_DIR}/var/run/slapd.args
+olcPidFile: ${LDAP_BASE_DIR}/var/run/slapd.pid
+
+#
+# Load dynamic backend modules commonly compiled in and available by default:
+#
+dn: cn=module,cn=config
+objectClass: olcModuleList
+cn: module
+olcModulePath: ${LDAP_BASE_DIR}/lib/openldap
+olcModuleLoad: back_mdb.la
+#olcModuleload: back_ldap.la
+#olcModuleload: back_passwd.la
+
+#
+# Schema settings
+#
+
+dn: cn=schema,cn=config
+objectClass: olcSchemaConfig
+cn: schema
+
+include: file://${LDAP_BASE_DIR}/etc/schema/core.ldif
+
+#
+# Frontend settings
+#
+
+dn: olcDatabase=frontend,cn=config
+objectClass: olcDatabaseConfig
+objectClass: olcFrontendConfig
+olcDatabase: frontend
+
+#
+# Configuration database
+#
+
+dn: olcDatabase=config,cn=config
+objectClass: olcDatabaseConfig
+olcDatabase: config
+olcAccess: to * by dn.base="gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth" manage by * none
+
+#
+# Server status monitoring
+#
+
+dn: olcDatabase=monitor,cn=config
+objectClass: olcDatabaseConfig
+olcDatabase: monitor
+olcAccess: to * by dn.base="gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth" read by dn.base="cn=Manager,dc=my-domain,dc=com" read by * none
+
+#
+# Backend database definitions
+#
+dn: olcDatabase=mdb,cn=config
+objectClass: olcDatabaseConfig
+objectClass: olcMdbConfig
+olcDatabase: mdb
+olcDbMaxSize: 1073741824
+olcSuffix: dc=my-domain,dc=com
+olcRootDN: cn=Manager,dc=my-domain,dc=com
+olcMonitoring: FALSE
+olcDbDirectory: ${LDAP_DATA_DIR}
```

Comment:
Not sure about putting the main DB in `${LDAP_DATA_DIR}` and others in subdirectories of that. No suggestions as to how it should be handled better.

Reply:
We could optionally provide an environment variable for a separate persistent volume where we could put the main DB.  When that's not set, we co-locate them, as is the case now.  Would that be better?

Reply:
It might be fine to have all as subdirs of LDAP_DATA_DIR for all this, volume wise that still gives people a choice to keep them separate that way. The main thing is not to have one db at "/path/*" and another at "/path/subdir/*".

Reply:
So, I'll move the `mdb` backend into `backend` a sub-dir of `/data`.  Is that a fair solution?

Reply:
Hmmm, that didn't work... why again is this layout not good?  How should it change?
```
/tmp/ldap
├── data
│   ├── accesslog
│   │   ├── data.mdb
│   │   └── lock.mdb
│   ├── data.mdb
│   └── lock.mdb
└── slapd.d
    ├── cn=config
    │   ├── cn=module{0}.ldif
    │   ├── cn=module{1}.ldif
    │   ├── cn=schema
    │   │   ├── cn={0}core.ldif
    │   │   ├── cn={1}cosine.ldif
    │   │   ├── cn={2}inetorgperson.ldif
    │   │   └── cn={3}nis.ldif
    │   ├── cn=schema.ldif
    │   ├── olcDatabase={0}config.ldif
    │   ├── olcDatabase={-1}frontend.ldif
    │   ├── olcDatabase={1}monitor.ldif
    │   ├── olcDatabase={2}mdb
    │   │   └── olcOverlay={0}accesslog.ldif
    │   ├── olcDatabase={2}mdb.ldif
    │   └── olcDatabase={3}mdb.ldif
    └── cn=config.ldif

6 directories, 18 files
```

Reply:
Fixed.
```
/tmp/ldap
├── data
│   ├── accesslog
│   │   ├── data.mdb
│   │   └── lock.mdb
│   └── backend
│       ├── data.mdb
│       └── lock.mdb
└── slapd.d
    ├── cn=config
    │   ├── cn=module{0}.ldif
    │   ├── cn=module{1}.ldif
    │   ├── cn=schema
    │   │   ├── cn={0}core.ldif
    │   │   ├── cn={1}cosine.ldif
    │   │   ├── cn={2}inetorgperson.ldif
    │   │   └── cn={3}nis.ldif
    │   ├── cn=schema.ldif
    │   ├── olcDatabase={0}config.ldif
    │   ├── olcDatabase={-1}frontend.ldif
    │   ├── olcDatabase={1}monitor.ldif
    │   ├── olcDatabase={2}mdb
    │   │   └── olcOverlay={0}accesslog.ldif
    │   ├── olcDatabase={2}mdb.ldif
    │   └── olcDatabase={3}mdb.ldif
    └── cn=config.ldif

7 directories, 18 files
```